includes/Products/class-fb-feed-generator.php

```diff
@@ -0,0 +1,579 @@
+	}
+
+	/**
+	 * Write data to the file.
+	 *
+	 * @since 3.1.0
+	 * @param string $data Data.
+	 */
+	protected function write_csv_data( $data ) {
+
+		if ( ! @file_exists( $this->get_file_path() ) ) { // phpcs:ignore Generic.PHP.NoSilencedErrors.Discouraged
+			@file_put_contents( $this->get_file_path(), $this->export_column_headers() ); // phpcs:ignore WordPress.VIP.FileSystemWritesDisallow.file_ops_file_put_contents, Generic.PHP.NoSilencedErrors.Discouraged, WordPress.WP.AlternativeFunctions.file_system_read_file_put_contents
+			@chmod( $this->get_file_path(), 0664 ); // phpcs:ignore WordPress.VIP.FileSystemWritesDisallow.chmod_chmod, WordPress.WP.AlternativeFunctions.file_system_read_file_put_contents, Generic.PHP.NoSilencedErrors.Discouraged
+		}
+		@file_put_contents( $this->get_file_path(), $data, FILE_APPEND ); // phpcs:ignore WordPress.VIP.FileSystemWritesDisallow.file_ops_file_put_contents, Generic.PHP.NoSilencedErrors.Discouraged, WordPress.WP.AlternativeFunctions.file_system_read_file_put_contents
+	}
+
+	public function execute_feed_generation_step() {
+		$settings = get_option( self::RUNNING_FEED_SETTINGS );
+		if ( $settings['done'] ) {
+			// We should not be here - add logging.
+			return;
+		}
+		$this->set_page( $settings['page'] );
+		$this->generate_file();
+		if ( ( $settings['page'] * $this->limit ) >= count( $settings['ids'] ) ) {
+			$settings['done'] = true;
+			$settings['end']  = time();
+			
+			update_option(
+				self::RUNNING_FEED_SETTINGS,
+				$settings,
+				false
+			);
+			$this->replace_feed_file_with_temp_file();
+			// Store all feed info so it will not get overwritten.
+			$file = $this->get_feed_directory() . $this->get_filename();
+			update_option(
+				self::FEED_FILE_INFO,
+				array(
+					'total'    => $settings['total'],
+					'start'    => $settings['start'],
+					'end'      => $settings['end'],
+					'location' => $file,
+					'size'     => round( filesize( $file ) / 1024 / 1024, 2 ),
+				)
+			);
+
+		} else {
+			$settings['page'] += 1;
+			update_option(
+				self::RUNNING_FEED_SETTINGS,
+				$settings,
+				false
+			);
+			as_enqueue_async_action( self::FEED_GENERATION_STEP );
+		}
+	}
+
+	/**
+	 * Last step of the feed generation procedure.
+	 * We have been writing to the temporary file until now.
+	 * We can safely delate old feed file and replace it with
+	 * the content of temporary file.
+	 */
+	public function replace_feed_file_with_temp_file() {
+		@rename(
+			$this->get_feed_directory() . $this->get_temp_filename(),
+			$this->get_feed_directory() . $this->get_filename()
+		); // phpcs:ignore WordPress.VIP.FileSystemWritesDisallow.file_ops_rename, Generic.PHP.NoSilencedErrors.Discouraged,
+	}
+
+	public function prepare_feed_generation() {
+		$this->prepare_feed_folder();
+		$products_ids = \WC_Facebookcommerce_Utils::get_all_product_ids_for_sync();
+		//$products_ids = array_slice( $products_ids, 0, 150, true );
+		update_option(
```

Comment:
In my experience with AW I've found reading and writing a large options value for every batch is slow and not as scalable as using querying items in chunks using an offset which is what I'm suggesting here: https://github.com/woocommerce/facebook-for-woocommerce/pull/1913/files#diff-441c6538f1e838f65093411744cb6e11279e449a05b9968ee30dea71ee82e288R50-R51

Reply:
Well, I think that this may be actually the opposite way 🤣 It all depends on the size of the items in the batch. Please see https://github.com/woocommerce/facebook-for-woocommerce/pull/1913/files#r622851965 comment.